core/src/main/scala/org/bitcoins/core/gcs/GolombFilter.scala

```diff
@@ -0,0 +1,147 @@
+package org.bitcoins.core.gcs
+
+import org.bitcoins.core.crypto.DoubleSha256Digest
+import org.bitcoins.core.number.{UInt64, UInt8}
+import org.bitcoins.core.protocol.CompactSizeUInt
+import org.bitcoins.core.protocol.blockchain.Block
+import org.bitcoins.core.protocol.script.{EmptyScriptPubKey, ScriptPubKey}
+import org.bitcoins.core.protocol.transaction.{
+  Transaction,
+  TransactionInput,
+  TransactionOutPoint,
+  TransactionOutput
+}
+import org.bitcoins.core.script.control.OP_RETURN
+import org.bitcoins.core.util.{BitcoinSUtil, BitcoinScriptUtil}
+import scodec.bits.{BitVector, ByteVector}
+
+import scala.annotation.tailrec
+
+// TODO: Replace ByteVector with a type for keys
+case class GolombFilter(
+    key: ByteVector,
+    m: UInt64,
+    p: UInt8,
+    n: CompactSizeUInt,
+    encodedData: BitVector) {
+  lazy val decodedHashes: Vector[UInt64] = GCS.golombDecodeSet(encodedData, p)
+
+  // TODO: Offer alternative that stops decoding when it finds out if data is there
+  def matchesHash(hash: UInt64): Boolean = {
+    @tailrec
+    def binarySearch(
+        from: Int,
+        to: Int,
+        hash: UInt64,
+        set: Vector[UInt64]): Boolean = {
+      if (to < from) {
+        false
+      } else {
+        val index = (to + from) / 2
+        val otherHash = set(index)
+
+        if (hash == otherHash) {
+          true
+        } else if (hash < otherHash) {
+          binarySearch(from, index - 1, hash, set)
+        } else {
+          binarySearch(index + 1, to, hash, set)
+        }
+      }
+    }
+
+    binarySearch(from = 0, to = n.toInt - 1, hash, decodedHashes)
+  }
+
+  def matches(data: ByteVector): Boolean = {
+    val f = n.num * m
+    val hash = GCS.hashToRange(data, f, key)
+
+    matchesHash(hash)
+  }
+}
+
+object BlockFilter {
+
+  val M: UInt64 = UInt64(784931)
+  val P: UInt8 = UInt8(19)
+
+  /**
+    * Given a Block and access to the UTXO set, constructs a Block Filter for that block
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters]]
+    */
+  def apply(block: Block, utxoProvider: TempUtxoProvider): GolombFilter = {
+    val key = block.blockHeader.hash.bytes.take(16)
+
+    val transactions: Vector[Transaction] = block.transactions.toVector
+
+    val noCoinbase: Vector[Transaction] = transactions.tail
+    val newOutputs: Vector[TransactionOutput] = transactions.flatMap(_.outputs)
+    val newScriptPubKeys: Vector[ByteVector] = newOutputs.flatMap { output =>
```

Comment:
This line took me a little while to figure out. But you're filtering here, right? If I understand this correctly, `None` and `Some` have implicit conversions available to `Vector`, which is why this work. What's your thoughts on using `.collect` instead? 

Also, I don't really see how the first if branch can be triggered. If we're dealing with an ampty SPK, it cannot contain OP_RETURN?

Reply:
Looking at your code in the method below, I'm suspecting that you meant to use `||` instead of `&&`